docs/codegen_passes.md

```diff
@@ -0,0 +1,572 @@
+      ...
+      %5 = subview %0[%arg0, %arg2]...
+      ...
+      %8 = subview %1[%arg2, %arg1]...
+      ...
+      %11 = subview %2[%arg0, %arg1]...
+      %12 = alloc(%c8, %c4) : memref<?x?xf32, 3>
+      %13 = subview %12[%c0, %c0]...
+      %14 = alloc(%c4, %c8) : memref<?x?xf32, 3>
+      %15 = subview %14[%c0, %c0]...
+      linalg.copy(%5, %13) {__internal_linalg_transform__ = "workgroup"}
+        : memref<?x?xf32, #map2>, memref<?x?xf32, #map2, 3>
+      spv.ControlBarrier "Workgroup", "Workgroup", "AcquireRelease"
+      linalg.copy(%8, %15) {__internal_linalg_transform__ = "workgroup"}
+        : memref<?x?xf32, #map2>, memref<?x?xf32, #map2, 3> spv.ControlBarrier
+      "Workgroup", "Workgroup", "AcquireRelease" linalg.matmul
+      {__internal_linalg_transform__ = "workgroup"} %13, %15, %11...
+      spv.ControlBarrier "Workgroup", "Workgroup", "AcquireRelease" dealloc
+      %12 : memref<?x?xf32, 3> dealloc %14 : memref<?x?xf32, 3> } scf.yield }
+      return } ``` Snippet 8: Tiling and promotion to use `Workgroup` memory for
+      `linalg.matmul`.  ``` <a name="snippet8"></a>Snippet 8: `linalg.matmul`
+      after tiling and promotion of operand subviews to use `Workgroup` memory.
+
+### Distributing to workgroups and workitems.
+
+After tiling the operations within the dispatch functions are either
+`scf.parallel` operations or Linalg operations.
+
+- The `scf.parallel` operations represent parallel loops that are to be
+  distributed across workgroups. The distribution here assumes that the number of workgroups along each dimension is equal to the number of iterations of the `scf.parallel`
+  operation.
+
+- Linalg operations that are not tiled, and are therefore __not within__ `scf`
+  operations, are lowered to loops. The resulting outer `scf.parallel` operations are
+  collapsed to have a single induction variable. This loop is then distributed
+  across workitems using their `GlobalInvocationId` (which is same as
+  `blockIdx.x * blockDim.x + threadIdx.x` in CUDA terminology).
+
+- Linalg operations that are tiled, and are therefore __within__ `scf` operations,
+  are lowered to loops and the iterations of the `scf.parallel` operations are
+  mapped to workitems using their `LocalInvocationId` (which is same as
+  `threadIdx.x` in CUDA terminology). Note that these operations are tagged with
+  the `workgroup` marker which makes it easy to disambiguate from the case where
+  Linalg operations are outside of `scf` operations. Here too, the distribution assumes that the workgroup size is greater than or equal to the number of
+  iterations of the partitioned loop.
+
+These transformations are applied by the [`ConvertToGPUPass`][ConvertToGPU]. Below is the result of applying this pass to Snippet 7. The outer `scf.parallel` loop is distributed across workgroups. The tiled `linalg.matmul` operation is lowered to loops, and the outer `scf.parallel` operation generated here are distributed across workitems within the workgroup.
```

Comment:
Nit: Wrap to 80 characters? (Other places too.) Otherwise it might cause trouble when importing/exporting to Google3.